Update decoding logic to handle special tokens

reachsumit opened the pull request:
## Description

Update decoding logic to handle special tokens

This change is targeted is a follow up to recently added `add_special_tokens` and `decode` features to `GPT2BPETokenizer`. 

The change adds the capability to decode special tokens ids back to text.

## Types of changes
[x ] New feature (non-breaking change which adds core functionality)

## Changes made
* Made changes to C++ code (`GPT2BPEEncoder` class) to update the `decode` functionality for special tokens.
* Added unit test to test the new decoding process.

## Testing
* No issue identified in `pre-commit`
* No issue identified with unit tests.

Nayef211 commented on torchtext/csrc/gpt2_bpe_tokenizer.cpp:
```diff
@@ -397,18 +397,45 @@ std::vector<int64_t> GPT2BPEEncoder::Encode(const std::string& text) {
+
+    // fix left space(s) for special tokens
+    if (special_token_flags[tok_idx] == true &&
+        (tok_idx > 0 && special_token_flags[tok_idx - 1] == false)) {
```
Just want to double check, does `&& special_token_flags[tok_idx - 1] == false` only ensure that we append a space if the token before the special token is a regular token? If so, why do we do this?

reachsumit replied:
It ensures that we insert space only if the last token wasn't special, because if the last token was special, it would have added space after it (through code at line 438). So we avoid having one extra space in case when two special tokens are next to each other.

Nayef211 reviewed (approved):
Overall changes LGTM. I would also wait for a stamp from @abhinavarora before merging this.

Nayef211 commented:
> @Nayef211 @reachsumit What is the status of this? Can we rebase, merge and let Jin's team know they have this capability now?

Changes looked fine to me but wanted a review from @abhinavarora as well to make sure I didn't miss anything!

reachsumit commented:
@joecummings I believe there is no additional PR required to support Jin's team's usecase. After this PR is approved and merged, we should be good to go.

abhinavarora commented on torchtext/csrc/gpt2_bpe_tokenizer.cpp:
```diff
@@ -397,18 +397,45 @@ std::vector<int64_t> GPT2BPEEncoder::Encode(const std::string& text) {
+      }
+    }
+
+    // fix left space(s) for special tokens
```
Let's add an example in comments to make this clearer?

abhinavarora commented on torchtext/csrc/gpt2_bpe_tokenizer.cpp:
```diff
@@ -397,18 +397,45 @@ std::vector<int64_t> GPT2BPEEncoder::Encode(const std::string& text) {
 
 std::string GPT2BPEEncoder::Decode(const std::vector<int64_t>& tokens) {
   std::string text;
+  std::vector<bool> special_token_flags(tokens.size());
```
Creating a vector seems to be an overkill. From the logic it seems we should be good with 2 bools. Is this correct or am I missing something?

abhinavarora reviewed (commented):
Left few comments. @reachsumit could you take a look? Logic looks fine to me.

reachsumit commented:
Rebased on latest `main` to fix conflicts, and addressed recent comments.
* Added examples for adding back spaces while decoding.
* Replaced using boolean array with two boolean variables.

abhinavarora commented on torchtext/csrc/gpt2_bpe_tokenizer.cpp:
```diff
@@ -400,19 +400,67 @@ std::vector<int64_t> GPT2BPEEncoder::Encode(const std::string& text) {
+          // get output character from byte decoder for each wide character
+          unsigned char uchr = byte_decoder_.at(converter.to_bytes(wchr));
+          decoded_token.push_back(uchr);
+          is_current_special = false;
```
How about setting this as `false` once at the beginning of the loop?

abhinavarora reviewed (approved):
Overall LGTM, left a minor comment.
